File: gulf_energy/gulf_energy/report/msds_expiry_report/msds_expiry_report.py
# Copyright (c) 2026, sammish and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import getdate, today, date_diff, add_days


def execute(filters=None):
	if not filters:
		filters = {}

	columns = get_columns()
	data = get_data(filters)

	return columns, data


def get_columns():
	return [
		{
			"label": _("MSDS ID"),
			"fieldname": "name",
			"fieldtype": "Link",
			"options": "MSDS Register",
			"width": 140,
		},
		{
			"label": _("Item Code"),
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 150,
		},
		{
			"label": _("Item Name"),
			"fieldname": "item_name",
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"label": _("Version"),
			"fieldname": "version",
			"fieldtype": "Data",
			"width": 80,
		},
		{
			"label": _("Status"),
			"fieldname": "status",
			"fieldtype": "Data",
			"width": 90,
		},
		{
			"label": _("Valid From"),
			"fieldname": "valid_from",
			"fieldtype": "Date",
			"width": 110,
		},
		{
			"label": _("Expiry Date"),
			"fieldname": "expiry_date",
			"fieldtype": "Date",
			"width": 110,
		},
		{
			"label": _("Days Remaining"),
			"fieldname": "days_remaining",
			"fieldtype": "Int",
			"width": 120,
		},
		{
			"label": _("Approved By"),
			"fieldname": "approved_by",
			"fieldtype": "Link",
			"options": "User",
			"width": 150,
		},
		{
			"label": _("Is Hazardous"),
			"fieldname": "is_hazardous",
			"fieldtype": "Check",
			"width": 100,
		},
		{
			"label": _("Remarks"),
			"fieldname": "remarks",
			"fieldtype": "Data",
			"width": 200,
		},
	]


def get_data(filters):
	conditions = ["1=1"]

	if filters.get("status"):
		conditions.append("msds.status = %(status)s")
	if filters.get("item_code"):
		conditions.append("msds.item_code = %(item_code)s")

	# Handle "expiring within days" filter
	expiry_filter = ""
	if filters.get("expiring_within_days"):
		days = int(filters["expiring_within_days"])
		cutoff_date = add_days(today(), days)
		expiry_filter = f"AND msds.expiry_date <= '{cutoff_date}' AND msds.expiry_date >= '{today()}'"
		# Only show Active records for expiry warning
		conditions.append("msds.status = 'Active'")

	where_clause = " AND ".join(conditions)

	data = frappe.db.sql(
		f"""
		SELECT
			msds.name,
			msds.item_code,
			msds.item_name,
			msds.version,
			msds.status,
			msds.valid_from,
			msds.expiry_date,
			msds.approved_by,
			msds.remarks,
			item.custom_is_hazardous as is_hazardous
		FROM
			`tabMSDS Register` msds
		LEFT JOIN
			`tabItem` item ON msds.item_code = item.name
		WHERE
			{where_clause}
			{expiry_filter}
		ORDER BY
			msds.expiry_date ASC
		""",
		filters,
		as_dict=1,
	)

	# Calculate days remaining
	today_date = getdate(today())
	for row in data:
		if row.expiry_date:
			row.days_remaining = date_diff(row.expiry_date, today_date)
		else:
			row.days_remaining = 0

	return data

File: gulf_energy/gulf_energy/report/msds_expiry_report/msds_expiry_report.js
// Copyright (c) 2026, sammish and contributors
// For license information, please see license.txt

frappe.query_reports["MSDS Expiry Report"] = {
	filters: [
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: ["", "Draft", "Active", "Expired"],
			default: "Active",
		},
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
		},
		{
			fieldname: "expiring_within_days",
			label: __("Expiring Within (Days)"),
			fieldtype: "Int",
			default: 30,
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "status") {
			if (data.status === "Active") {
				value = `<span class="text-success font-weight-bold">${value}</span>`;
			} else if (data.status === "Expired") {
				value = `<span class="text-danger font-weight-bold">${value}</span>`;
			} else if (data.status === "Draft") {
				value = `<span class="text-warning">${value}</span>`;
			}
		}

		if (column.fieldname === "days_remaining") {
			if (data.days_remaining <= 0) {
				value = `<span class="text-danger font-weight-bold">${value}</span>`;
			} else if (data.days_remaining <= 30) {
				value = `<span class="text-warning font-weight-bold">${value}</span>`;
			} else {
				value = `<span class="text-success">${value}</span>`;
			}
		}

		if (column.fieldname === "is_hazardous" && data.is_hazardous) {
			value = `<span class="text-danger font-weight-bold">Yes</span>`;
		}

		return value;
	},
};
